crates/ethrpc/src/instrumented.rs

```diff
@@ -0,0 +1,148 @@
+use {
+    ethcontract::{
+        dyns::DynWeb3,
+        jsonrpc::types::{Call, Value},
+        transport::DynTransport,
+    },
+    futures::{future::BoxFuture, FutureExt},
+    std::sync::Arc,
+    web3::{error::Error as Web3Error, BatchTransport, RequestId, Transport},
+};
+
+#[derive(prometheus_metric_storage::MetricStorage, Clone, Debug)]
+#[metric(subsystem = "rpc")]
+struct Metrics {
+    /// Number of inflight RPC requests for ethereum node.
+    #[metric(labels("component", "method"))]
+    requests_inflight: prometheus::IntGaugeVec,
+
+    /// Number of completed RPC requests for ethereum node.
+    #[metric(labels("component", "method"))]
+    requests_complete: prometheus::IntCounterVec,
+
+    /// Execution time for each RPC request (batches are counted as one
+    /// request).
+    #[metric(labels("component", "method"))]
+    requests_duration_seconds: prometheus::HistogramVec,
+
+    /// Number of RPC requests initiated within a batch request
+    #[metric(labels("component", "method"))]
+    inner_batch_requests_initiated: prometheus::IntCounterVec,
+}
+
+impl Metrics {
+    #[must_use]
+    fn on_request_start(&self, label: &str, method: &str) -> impl Drop {
+        let requests_inflight = self.requests_inflight.with_label_values(&[label, method]);
+        let requests_complete = self.requests_complete.with_label_values(&[label, method]);
+        let requests_duration_seconds = self
+            .requests_duration_seconds
+            .with_label_values(&[label, method]);
+
+        requests_inflight.inc();
+        let timer = requests_duration_seconds.start_timer();
+
+        scopeguard::guard(timer, move |timer| {
+            requests_inflight.dec();
+            requests_complete.inc();
+            timer.stop_and_record();
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct InstrumentedTransport(Arc<Inner>);
+
+impl InstrumentedTransport {
+    pub fn new(label: String, transport: DynTransport) -> Self {
+        Self(Arc::new(Inner {
+            metrics: Metrics::instance(observe::metrics::get_storage_registry()).unwrap(),
+            transport,
+            label,
+        }))
+    }
+
+    pub fn with_additional_label(&self, label: String) -> Self {
+        Self(Arc::new(Inner {
+            label: format!("{}_{label}", self.0.label),
+            transport: self.0.transport.clone(),
```

Comment:
Is this doing a shallow copy in that we can continue the original transport (e.g. for batching purposes). Otherise, do we need to wrap this in an Arc here?

Reply:
Depends on the `Transport` but we are only using the `BufferedTransport` and `HttpTransport` both of which implement shallow copies.
For `BufferedTransport` that means we continue to use the same batching structure and `HttpTransport` keeps using the same counter for trace logs and such.

Technically it would probably be safer to use `Arc` just so we don't have to worry about the actual `Clone` implementation of the `Transport` in question but then I would be slightly worried about all the pointer chasing we have to do for every single request.